magnum-cats-effect/src/main/scala/com/augustnagro/magnum/magcats/Transactor.scala

```diff
@@ -0,0 +1,202 @@
+package com.augustnagro.magnum.magcats
+
+import cats.*
+import cats.effect.kernel.*
+import cats.effect.kernel.Outcome.*
+import cats.effect.std.*
+import cats.effect.syntax.all.*
+import cats.syntax.all.*
+import com.augustnagro.magnum.*
+
+import java.sql.Connection
+import javax.sql.DataSource
+import scala.util.control.NonFatal
+
+class Transactor[F[_]: Sync] private (
+    dataSource: DataSource,
+    sqlLogger: SqlLogger,
+    connectionConfig: Connection => Unit,
+    rateLimiter: Option[Resource[F, Unit]]
+):
+  private val makeConn = Resource.make(acquireConnection)(releaseConnection)
+
+  def withSqlLogger(sqlLogger: SqlLogger): Transactor[F] =
+    new Transactor(
+      dataSource,
+      sqlLogger,
+      connectionConfig,
+      rateLimiter
+    )
+
+  def withConnectionConfig(
+      connectionConfig: Connection => Unit
+  ): Transactor[F] =
+    new Transactor(
+      dataSource,
+      sqlLogger,
+      connectionConfig,
+      rateLimiter
+    )
+
+  def connect[A](f: DbCon ?=> A): F[A] =
+    useRateLimitedConnection: cn =>
+      Sync[F].delay(connectionConfig(cn)) >>
+        Sync[F].interruptible(f(using DbCon(cn, sqlLogger)))
+
+  def transact[A](f: DbTx ?=> A): F[A] =
+    useRateLimitedConnection: cn =>
+      Sync[F].delay(connectionConfig(cn)) >>
+        Sync[F].delay(cn.setAutoCommit(false)) >>
```

Comment:
To get a slightly more performant code, you could wrap these 2 lines of code into only 1 `.delay` call

That's the reason the ZIO integration wraps all this code into only 1 `ZIO.attempt { ... }`